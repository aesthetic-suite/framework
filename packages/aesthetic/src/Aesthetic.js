/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { isObject } from 'aesthetic-utils';
import deepMerge from 'lodash.merge';
import Adapter from './Adapter';

import type {
  AestheticOptions,
  ClassName,
  GlobalDeclaration,
  StyleCallback,
  Statement,
  ThemeDeclaration,
  StyleSheet,
} from '../../types';

export default class Aesthetic {
  adapter: Adapter;

  cache: { [styleName: string]: StyleSheet } = {};

  native: boolean = false;

  options: AestheticOptions = {
    defaultTheme: '',
    extendable: false,
    pure: false,
    stylesPropName: 'classNames',
    themePropName: 'theme',
  };

  parents: { [childStyleName: string]: string } = {};

  styles: { [styleName: string]: StyleCallback | Statement } = {};

  themes: { [themeName: string]: ThemeDeclaration } = {};

  constructor(adapter: Adapter, options?: Object = {}) {
    this.options = {
      ...this.options,
      ...options,
    };

    this.setAdapter(adapter);
  }

  /**
   * Register a theme by extending and merging with a previously defined theme.
   */
  extendTheme(
    parentThemeName: string,
    themeName: string,
    theme?: ThemeDeclaration = {},
    globals?: GlobalDeclaration = {},
  ): this {
    return this.registerTheme(
      themeName,
      deepMerge({}, this.getTheme(parentThemeName), theme),
      globals,
    );
  }

  /**
   * Extract the defined style declarations. If the declaratin is a function,
   * execute it while passing the current theme and previous inherited styles.
   */
  getStyles(styleName: string, themeName?: string = ''): Statement {
    const parentStyleName = this.parents[styleName];
    const declarations = this.styles[styleName];

    if (__DEV__) {
      if (!declarations) {
        throw new Error(`Styles do not exist for "${styleName}".`);
      }
    }

    if (typeof declarations !== 'function') {
      return declarations;
    }

    return declarations(
      themeName ? this.getTheme(themeName) : {},
      parentStyleName ? this.getStyles(parentStyleName, themeName) : {},
    );
  }

  /**
   * Return a themes style object or throw an error.
   */
  getTheme(themeName?: string = ''): ThemeDeclaration {
    const { defaultTheme } = this.options;

    let theme = this.themes[themeName];

    if (!theme && defaultTheme) {
      theme = this.themes[defaultTheme];
    }

    if (__DEV__) {
      if (!theme) {
        throw new Error(`Theme "${themeName}" does not exist.`);
      }
    }

    return theme;
  }

  /**
   * Register a theme with a pre-defined set of theme settings.
   */
  registerTheme(
    themeName: string,
    theme?: ThemeDeclaration = {},
    globals?: GlobalDeclaration = {},
  ): this {
    if (__DEV__) {
      if (this.themes[themeName]) {
        throw new Error(`Theme "${themeName}" already exists.`);

      } else if (!isObject(theme)) {
        throw new TypeError(`Theme "${themeName}" must be a style object.`);

      } else if (!isObject(globals)) {
        throw new TypeError(`Global styles for "${themeName}" must be an object.`);
      }
    }

    // Register the theme
    this.themes[themeName] = theme;

    // Transform the global styles
    this.adapter.transform(':root', globals);

    return this;
  }

  /**
   * Set an adapter class to transform CSS style objects.
   */
  setAdapter(adapter: Adapter): this {
    if (adapter instanceof Adapter || (adapter && typeof adapter.transform === 'function')) {
      adapter.native = this.native; // eslint-disable-line
      this.adapter = adapter;

    } else if (__DEV__) {
      throw new TypeError('Adapter must be an instance of `Adapter`.');
    }

    return this;
  }

  /**
   * Set multiple style declarations for a component.
   */
  setStyles(
    styleName: string,
    statement: StyleCallback | Statement,
    extendFrom?: string = '',
  ): this {
    if (__DEV__) {
      if (this.styles[styleName]) {
        throw new Error(`Styles have already been set for "${styleName}".`);

      } else if (!isObject(statement) && typeof statement !== 'function') {
        throw new TypeError(`Styles defined for "${styleName}" must be an object or function.`);
      }
    }

    this.styles[styleName] = statement;

    if (extendFrom) {
      if (__DEV__) {
        if (!this.styles[extendFrom]) {
          throw new Error(`Cannot extend from "${extendFrom}" as those styles do not exist.`);

        } else if (extendFrom === styleName) {
          throw new Error('Cannot extend styles from itself.');
        }
      }

      this.parents[styleName] = extendFrom;
    }

    return this;
  }

  /**
   * Execute the adapter transformer on the set of style declarations for the
   * defined component. Optionally support a custom theme.
   */
  transformStyles(styleName: string, themeName?: string): StyleSheet {
    const fallbackThemeName = themeName || this.options.defaultTheme || '';
    const cacheKey = `${styleName}:${fallbackThemeName}`;

    if (this.cache[cacheKey]) {
      return this.cache[cacheKey];
    }

    const declarations = this.getStyles(styleName, fallbackThemeName);
    const toTransform = {};
    const output = {};
    let setCount = 0;

    // Separate style objects from class names
    Object.keys(declarations).forEach((selector) => {
      if (typeof declarations[selector] === 'string') {
        output[selector] = this.native ? {} : declarations[selector];
      } else {
        toTransform[selector] = declarations[selector];
        setCount += 1;
      }
    });

    // Transform the styles into a map of class names
    if (setCount > 0) {
      const transformedOutput = this.adapter.transform(styleName, toTransform);

      Object.keys(transformedOutput).forEach((selector) => {
        output[selector] = this.validateTransform(styleName, selector, transformedOutput[selector]);
      });
    }

    // Cache the values
    this.cache[cacheKey] = output;

    return output;
  }

  /**
   * Validate the object returned contains valid strings.
   */
  validateTransform(styleName: string, selector: string, value: ClassName): ClassName {
    if (__DEV__) {
      if (typeof value !== 'string') {
        throw new TypeError(
          `\`${this.adapter.constructor.name}\` must return a mapping of CSS class names. ` +
          `"${styleName}@${selector}" is not a valid string.`,
        );
      }
    }

    return value;
  }
}
