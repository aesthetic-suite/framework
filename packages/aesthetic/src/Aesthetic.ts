/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import isObject from './helpers/isObject';
import stripClassPrefix from './helpers/stripClassPrefix';
import Adapter from './Adapter';
import withStyles from './style';
import { ClassName, $FixMe, StyleName, ThemeName } from './types';

export interface AestheticOptions {
  defaultTheme: ThemeName;
  extendable: boolean;
  passThemeNameProp: boolean;
  passThemeProp: boolean;
  pure: boolean;
  stylesPropName: string;
  themePropName: string;
}

export default class Aesthetic<Theme, StyleSheet, Declaration> {
  adapter: Adapter<StyleSheet, Declaration>;

  cache: WeakMap<Declaration[], ClassName> = new WeakMap();

  options: AestheticOptions = {
    defaultTheme: '',
    extendable: false,
    passThemeNameProp: true,
    passThemeProp: true,
    pure: false,
    stylesPropName: 'styles',
    themePropName: 'theme',
  };

  parents: { [childStyleName: string]: string } = {};

  styles: { [styleName: string]: StyleSheet | StyleSheetCallback } = {};

  themes: { [themeName: string]: Theme } = {};

  constructor(adapter: Adapter<StyleSheet, Declaration>, options: Partial<AestheticOptions> = {}) {
    this.adapter = adapter;
    this.options = {
      ...this.options,
      ...options,
    };
  }

  /**
   * Return a stylesheet unique to an adapter.
   */
  createStyleSheet(
    styleName: StyleName,
    themeName: ThemeName = '',
    props: $FixMe = {},
  ): StyleSheet {
    return this.adapter.create(this.getStyles(styleName, themeName, props), styleName);
  }

  /**
   * Register a theme by extending and merging with a previously defined theme.
   */
  extendTheme(
    parentThemeName: ThemeName,
    themeName: ThemeName,
    theme: Theme,
    globals?: $FixMe,
  ): this {
    return this.registerTheme(
      themeName,
      this.adapter.merge(this.getTheme(parentThemeName), theme),
      globals,
    );
  }

  /**
   * Retrieve the defined style declarations. If the declaratin is a function,
   * execute it while passing the current theme and React props.
   */
  getStyles(styleName: StyleName, themeName: ThemeName = '', props: $FixMe = {}): StyleSheet {
    const parentStyleName = this.parents[styleName];
    let styleSheet = this.styles[styleName];

    if (process.env.NODE_ENV !== 'production') {
      if (!styleSheet) {
        throw new Error(`Styles do not exist for "${styleName}".`);
      }
    }

    // Extract styleSheet from callback
    if (typeof styleSheet === 'function') {
      styleSheet = styleSheet(themeName ? this.getTheme(themeName) : {}, props);
    }

    // Merge from parent
    if (parentStyleName) {
      styleSheet = this.adapter.merge(
        this.getStyles(parentStyleName, themeName, props),
        styleSheet,
      );
    }

    return styleSheet;
  }

  /**
   * Return a themes style object or throw an error.
   */
  getTheme(themeName: ThemeName = ''): Theme {
    const { defaultTheme } = this.options;

    let theme = this.themes[themeName];

    if (!theme && defaultTheme) {
      theme = this.themes[defaultTheme];
    }

    if (process.env.NODE_ENV !== 'production') {
      if (!theme) {
        throw new Error(`Theme "${themeName}" does not exist.`);
      }
    }

    return theme;
  }

  /**
   * Register a theme with a pre-defined set of theme settings.
   */
  registerTheme(themeName: ThemeName, theme: Theme, globals?: StyleSheet): this {
    if (process.env.NODE_ENV !== 'production') {
      if (this.themes[themeName]) {
        throw new Error(`Theme "${themeName}" already exists.`);
      } else if (!isObject(theme)) {
        throw new TypeError(`Theme "${themeName}" must be a style object.`);
      } else if (!isObject(globals)) {
        throw new TypeError(`Global styles for "${themeName}" must be an object.`);
      }
    }

    // Register the theme
    this.themes[themeName] = theme;

    // Create global styles
    if (globals) {
      const globalStyleSheet = this.adapter.create(globals, ':root');

      this.transformStyles(Object.values(globalStyleSheet));
    }

    return this;
  }

  /**
   * Set multiple style declarations for a component.
   */
  setStyles(
    styleName: StyleName,
    styleSheet: StyleSheet | StyleSheetCallback,
    extendFrom: StyleName = '',
  ): this {
    if (process.env.NODE_ENV !== 'production') {
      if (this.styles[styleName]) {
        throw new Error(`Styles have already been set for "${styleName}".`);
      } else if (!isObject(styleSheet) && typeof styleSheet !== 'function') {
        throw new TypeError(`Styles defined for "${styleName}" must be an object or function.`);
      }
    }

    this.styles[styleName] = styleSheet;

    if (extendFrom) {
      if (process.env.NODE_ENV !== 'production') {
        if (!this.styles[extendFrom]) {
          throw new Error(`Cannot extend from "${extendFrom}" as those styles do not exist.`);
        } else if (extendFrom === styleName) {
          throw new Error('Cannot extend styles from itself.');
        }
      }

      this.parents[styleName] = extendFrom;
    }

    return this;
  }

  /**
   * Execute the adapter transformer on the list of style declarations.
   */
  transformStyles(styles: Declaration[]): ClassName {
    if (this.cache.has(styles)) {
      return this.cache.get(styles)!;
    }

    const classNames: ClassName[] = [];
    const toTransform: Declaration[] = [];

    styles.forEach(style => {
      // Empty value or failed condition
      if (!style) {
        return; // eslint-disable-line

        // Acceptable class names
      } else if (typeof style === 'string' || typeof style === 'number') {
        classNames.push(
          ...String(style)
            .split(' ')
            .map(s => stripClassPrefix(s).trim()),
        );

        // Style objects
      } else if (isObject(style)) {
        toTransform.push(style);
      } else if (process.env.NODE_ENV !== 'production') {
        throw new Error('Unsupported style type to transform.');
      }
    });

    if (toTransform.length > 0) {
      classNames.push(this.adapter.transform(...toTransform));
    }

    const className = classNames.join(' ').trim();

    this.cache.set(styles, className);

    return className;
  }

  /**
   * Utility method for wrapping a component with a styles HOC.
   */
  withStyles(styleSheet: StyleSheet | StyleSheetCallback, options?: HOCOptions = {}): HOCWrapper {
    return withStyles(this, styleSheet, options);
  }
}
