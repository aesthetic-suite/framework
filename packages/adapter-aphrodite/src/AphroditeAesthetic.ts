import Aesthetic, { AestheticOptions, ClassName, Ruleset, Sheet, SheetMap } from 'aesthetic';
// @ts-ignore flushToStyleTag is not typed
import { StyleSheet as Aphrodite, Extension, flushToStyleTag } from 'aphrodite';
import { NativeBlock, ParsedBlock } from './types';

export default class AphroditeAesthetic<Theme extends object> extends Aesthetic<
  Theme,
  NativeBlock,
  ParsedBlock
> {
  aphrodite: {
    StyleSheet: typeof Aphrodite;
    css(...styles: ParsedBlock[]): ClassName;
  };

  fontFaces: { [fontFamily: string]: NativeBlock[] } = {};

  keyframes: { [animationName: string]: NativeBlock } = {};

  constructor(extensions: Extension[] = [], options: Partial<AestheticOptions> = {}) {
    super(options);

    this.aphrodite = Aphrodite.extend([
      ...extensions,
      { selectorHandler: this.handleHierarchySelector },
      { selectorHandler: this.handleGlobalSelector },
    ]);

    this.syntax
      .on('attribute', this.handleNested)
      .on('css', this.handleCss)
      .on('font-face', this.handleFontFace)
      .on('global', this.handleGlobal)
      .on('keyframe', this.handleKeyframe)
      .on('media', this.handleMedia)
      .on('property', this.handleProperty)
      .on('pseudo', this.handleNested)
      .on('selector', this.handleNested);
  }

  flushStyles() {
    flushToStyleTag();
  }

  isParsedBlock(block: NativeBlock | ParsedBlock): block is ParsedBlock {
    // eslint-disable-next-line no-underscore-dangle
    return Boolean(block && block._name && block._definition);
  }

  protected processStyleSheet(styleSheet: SheetMap<NativeBlock>): SheetMap<ParsedBlock> {
    return this.aphrodite.StyleSheet.create(styleSheet) as SheetMap<ParsedBlock>;
  }

  protected transformToClassName(styles: ParsedBlock[]): ClassName {
    return this.aphrodite.css(...styles);
  }

  private handleCss = (css: string) => {
    this.getStyleSheetManager().injectStatements(css);
  };

  // https://github.com/Khan/aphrodite#font-faces
  private handleFontFace = (
    sheet: Sheet<NativeBlock>,
    fontFaces: Ruleset<NativeBlock>[],
    fontFamily: string,
  ) => {
    this.fontFaces[fontFamily] = fontFaces.map(face => face.toObject());
  };

  private handleGlobal = (
    sheet: Sheet<NativeBlock>,
    selector: string,
    ruleset: Ruleset<NativeBlock>,
  ) => {
    const current: Ruleset<NativeBlock> = sheet.ruleSets.globals || sheet.createRuleset('globals');

    current.addNested(`*${selector}`, ruleset);

    sheet.addRuleset(current);
  };

  private handleGlobalSelector(
    selector: string,
    baseSelector: string,
    callback: (selector: string) => string,
  ): string | null {
    if (selector.charAt(0) !== '*') {
      return null;
    }

    return callback(selector.slice(1));
  }

  private handleHierarchySelector(
    selector: string,
    baseSelector: string,
    callback: (selector: string) => string,
  ): string | null {
    if (selector.charAt(0) === '>') {
      return callback(`${baseSelector} ${selector}`);
    }

    if (selector.charAt(0) === '[') {
      return callback(`${baseSelector}${selector}`);
    }

    return null;
  }

  // https://github.com/Khan/aphrodite#animations
  private handleKeyframe = (
    sheet: Sheet<NativeBlock>,
    keyframe: Ruleset<NativeBlock>,
    animationName: string,
  ) => {
    this.keyframes[animationName] = keyframe.toObject();
  };

  // https://github.com/Khan/aphrodite#api
  private handleMedia = (
    ruleset: Ruleset<NativeBlock>,
    query: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(`@media ${query}`, value);
  };

  // https://github.com/Khan/aphrodite#api
  private handleNested = (
    ruleset: Ruleset<NativeBlock>,
    selector: string,
    value: Ruleset<NativeBlock>,
  ) => {
    ruleset.addNested(selector, value);
  };

  // https://github.com/Khan/aphrodite#api
  private handleProperty = (ruleset: Ruleset<NativeBlock>, name: keyof NativeBlock, value: any) => {
    if (name === 'animationName') {
      ruleset.addProperty(name, this.syntax.injectKeyframes(value, this.keyframes));
    } else if (name === 'fontFamily') {
      // Font faces could potentially convert recursively because font faces
      // have a `familyName`, and we parse on `familyName`. Luckily the ruleset
      // selector is the family name, and we can determine that this is being
      // called from from `convertFontFaces`.
      if (value.includes(ruleset.selector)) {
        ruleset.addProperty(name, value);
      } else {
        ruleset.addProperty(name, this.syntax.injectFontFaces(value, this.fontFaces));
      }
    } else {
      ruleset.addProperty(name, value);
    }
  };
}
