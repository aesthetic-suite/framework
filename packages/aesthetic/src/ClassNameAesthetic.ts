/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Aesthetic from './Aesthetic';
import { ClassName } from './types';

export default class ClassNameAesthetic<Theme> extends Aesthetic<Theme, ClassName> {
  transformToClassName(...styles: ClassName[]): ClassName {
    const classNames: ClassName[] = [];

    styles.forEach(style => {
      if (style && typeof style === 'string') {
        classNames.push(style);
      } else if (process.env.NODE_ENV !== 'production') {
        throw new TypeError(`${this.constructor.name} expects valid CSS class names.`);
      }
    });

    return classNames.join(' ');
  }
}
