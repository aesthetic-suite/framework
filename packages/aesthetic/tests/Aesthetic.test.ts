/* eslint-disable sort-keys */

import { StyleSheetTestUtils } from 'aphrodite';
import { createRenderer } from 'fela';
import webPreset from 'fela-preset-web';
import { create } from 'jss';
import preset from 'jss-preset-default';
import Aesthetic from '../src/Aesthetic';
import ClassNameAesthetic from '../src/ClassNameAesthetic';
import AphroditeAesthetic from '../../aesthetic-adapter-aphrodite/src';
import CSSModulesAesthetic from '../../aesthetic-adapter-css-modules/src';
import FelaAesthetic from '../../aesthetic-adapter-fela/src';
import JSSAesthetic from '../../aesthetic-adapter-jss/src';
import TypeStyleAesthetic from '../../aesthetic-adapter-typestyle/src';
import { SYNTAX_GLOBAL, SYNTAX_UNIFIED_LOCAL_FULL } from '../../../tests/mocks';

describe('Aesthetic', () => {
  let instance: Aesthetic<any, any, any>;

  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();

    instance = new ClassNameAesthetic();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  describe('constructor()', () => {
    it('merges options', () => {
      instance = new ClassNameAesthetic({
        stylesPropName: 'styleSheet',
      });

      expect(instance.options).toEqual({
        extendable: false,
        passThemeNameProp: true,
        passThemeProp: true,
        pure: false,
        stylesPropName: 'styleSheet',
        theme: 'default',
        themePropName: 'theme',
      });
    });
  });

  describe('createStyleSheet()', () => {
    it('returns the style sheet', () => {
      instance.themes.classic = {};
      instance.styles.foo = () => ({
        el: { display: 'block' },
      });

      // Will have no properties as no syntax handlers are defined
      expect(instance.createStyleSheet('foo', 'classic')).toEqual({
        el: {},
      });
    });
  });

  describe('extendTheme()', () => {
    it('errors if the parent theme doesnt exist', () => {
      expect(() => instance.extendTheme('foo', 'bar', {})).toThrowErrorMatchingSnapshot();
    });

    it('deep merges the parent and child theme', () => {
      instance.themes.foo = {
        unit: 'px',
        unitSize: 8,
        colors: {
          primary: 'red',
        },
      };

      instance.extendTheme('foo', 'bar', {
        unit: 'em',
        colors: {
          primary: 'blue',
          secondary: 'orange',
        },
      });

      expect(instance.themes.bar).toEqual({
        unit: 'em',
        unitSize: 8,
        colors: {
          primary: 'blue',
          secondary: 'orange',
        },
      });
    });
  });

  describe('getStyles()', () => {
    beforeEach(() => {
      instance.themes.classic = {};
    });

    it('errors if no theme', () => {
      instance.styles.foo = () => ({
        el: { display: 'block' },
      });

      expect(() => instance.getStyles('foo', 'unknown')).toThrowErrorMatchingSnapshot();
    });

    it('returns an empty object if null', () => {
      instance.styles.foo = null;

      expect(instance.getStyles('foo', 'classic')).toEqual({});
    });

    it('returns the stylesheet', () => {
      instance.styles.foo = () => ({
        el: { display: 'block' },
      });

      expect(instance.getStyles('foo', 'classic')).toEqual({
        el: { display: 'block' },
      });
    });

    it('passes theme to the style callback', () => {
      instance.themes.classic = {
        unitSize: 5,
      };

      instance.styles.foo = theme => ({
        el: { padding: theme.unitSize * 2 },
      });

      expect(instance.getStyles('foo', 'classic')).toEqual({
        el: { padding: 10 },
      });
    });

    it('passes props to the style callback', () => {
      instance.styles.foo = (theme, props) => ({
        el: { padding: props.unitSize * 2 },
      });

      expect(
        instance.getStyles('foo', 'classic', {
          unitSize: 5,
        }),
      ).toEqual({
        el: { padding: 10 },
      });
    });

    it('inherits styles from parent', () => {
      instance.setStyles('foo', () => ({
        el: {
          color: 'red',
          ':hover': {
            color: 'darkred',
          },
        },
      }));

      instance.setStyles(
        'bar',
        () => ({
          el: {
            background: 'blue',
            ':hover': {
              color: 'green',
            },
          },
        }),
        'foo',
      );

      instance.setStyles(
        'baz',
        () => ({
          el: { display: 'block' },
        }),
        'bar',
      );

      expect(instance.getStyles('foo', 'classic')).toEqual({
        el: {
          color: 'red',
          ':hover': {
            color: 'darkred',
          },
        },
      });

      expect(instance.getStyles('bar', 'classic')).toEqual({
        el: {
          color: 'red',
          background: 'blue',
          ':hover': {
            color: 'green',
          },
        },
      });

      expect(instance.getStyles('baz', 'classic')).toEqual({
        el: {
          color: 'red',
          background: 'blue',
          display: 'block',
          ':hover': {
            color: 'green',
          },
        },
      });
    });
  });

  describe('getTheme()', () => {
    it('errors if the theme doesnt exist', () => {
      expect(() => instance.getTheme('foo')).toThrowErrorMatchingSnapshot();
    });

    it('returns the theme by name', () => {
      instance.registerTheme('foo', { unitSize: 6 });

      expect(instance.getTheme('foo')).toEqual({ unitSize: 6 });
    });

    it('returns the default theme if defined and requested them doesnt exist', () => {
      instance.registerTheme('default', { unitSize: 1 });
      instance.registerTheme('foo', { unitSize: 6 });
      instance.options.defaultTheme = 'default';

      expect(instance.getTheme('bar')).toEqual({ unitSize: 1 });
    });
  });

  describe('processStyleSheet()', () => {
    it('returns the styleSheet as a stylesheet', () => {
      const styleSheet = { el: {} };
      const stylesheet = instance.processStyleSheet(styleSheet, 'styleName');

      expect(styleSheet).toEqual(stylesheet);
    });
  });

  describe('registerTheme()', () => {
    it('errors if a theme name has been used', () => {
      instance.themes.foo = {};

      expect(() => instance.registerTheme('foo', {})).toThrowErrorMatchingSnapshot();
    });

    it('errors if a theme style is not an object', () => {
      expect(() => instance.registerTheme('foo', 123)).toThrowErrorMatchingSnapshot();
    });

    it('errors if global styles is not an object', () => {
      // @ts-ignore Allow non-object
      expect(() => instance.registerTheme('foo', {}, 123)).toThrowErrorMatchingSnapshot();
    });

    it('registers theme and sets global styles', () => {
      expect(instance.globals.foo).toBeUndefined();

      instance.registerTheme('foo', { unitSize: 6 }, () => SYNTAX_GLOBAL);

      expect(instance.themes).toEqual({
        foo: { unitSize: 6 },
      });

      expect(instance.globals.foo).toBeDefined();
    });
  });

  describe('setStyles()', () => {
    it('errors if styles have been set', () => {
      instance.styles.foo = () => ({});

      expect(() => instance.setStyles('foo', null)).toThrowErrorMatchingSnapshot();
    });

    it('errors if styles try to extend from itself', () => {
      instance.styles.foo = () => ({});

      expect(() => instance.setStyles('foo', () => ({}), 'foo')).toThrowErrorMatchingSnapshot();
    });

    it('errors if styles are not a function', () => {
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyles('foo', 123)).toThrowErrorMatchingSnapshot();
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyles('foo', 'abc')).toThrowErrorMatchingSnapshot();
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyles('foo', [])).toThrowErrorMatchingSnapshot();
      // @ts-ignore Allow invalid type
      expect(() => instance.setStyles('foo', true)).toThrowErrorMatchingSnapshot();
    });

    it('errors if extended styles do not exist', () => {
      expect(() => instance.setStyles('foo', null, 'parent')).toThrowErrorMatchingSnapshot();
    });

    it('errors if extended and style names match', () => {
      expect(() => instance.setStyles('foo', null, 'foo')).toThrowErrorMatchingSnapshot();
    });

    it('sets styles', () => {
      expect(instance.styles.foo).toBeUndefined();

      instance.setStyles('foo', () => ({
        header: { color: 'red' },
        footer: { padding: 5 },
      }));

      expect(instance.styles.foo).toBeDefined();
    });

    it('sets styles and extends parents', () => {
      expect(instance.styles.bar).toBeUndefined();
      expect(instance.parents.bar).toBeUndefined();

      instance.setStyles('foo', () => ({
        header: { color: 'red' },
        footer: { padding: 5 },
      }));

      instance.setStyles(
        'bar',
        () => ({
          child: { margin: 5 },
        }),
        'foo',
      );

      expect(instance.styles.bar!({}, {})).toEqual({
        child: { margin: 5 },
      });
      expect(instance.parents.bar).toBe('foo');
    });
  });

  describe('transformStyles()', () => {
    beforeEach(() => {
      instance = new AphroditeAesthetic();
    });

    it('errors for invalid value', () => {
      expect(() => {
        instance.transformStyles([[]]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('combines strings into a class name', () => {
      expect(instance.transformStyles('foo', 'bar')).toBe('foo bar');
    });

    it('calls transformToClassName() method', () => {
      const spy = jest.fn();

      instance.transformToClassName = spy;
      instance.transformStyles({ color: 'red' }, { display: 'block' });

      expect(spy).toHaveBeenCalledWith([{ color: 'red' }, { display: 'block' }]);
    });

    it('ignores falsey values', () => {
      expect(instance.transformStyles(null, false, 0, '', undefined)).toBe('');
    });

    it('strips period prefix', () => {
      expect(instance.transformStyles('.foo', 'bar .qux')).toBe('foo bar qux');
    });

    it('handles expression values', () => {
      expect(instance.transformStyles('foo', true && 'bar', 5 > 10 && 'baz')).toBe('foo bar');
    });

    it('joins strings and numbers', () => {
      expect(instance.transformStyles('foo', 123, 'bar')).toBe('foo 123 bar');
    });
  });

  describe('adapters', () => {
    it('supports class names', () => {
      instance = new ClassNameAesthetic();
      instance.registerTheme('default', {});
      instance.setStyles('foo', () => ({ button: 'button' }));

      const styleSheet = instance.createStyleSheet('foo', 'default');

      expect(instance.transformStyles(styleSheet.button)).toMatchSnapshot();
    });

    it('supports Aphrodite', () => {
      instance = new AphroditeAesthetic();
      instance.registerTheme('default', {});
      instance.setStyles('foo', () => SYNTAX_UNIFIED_LOCAL_FULL);

      const styleSheet = instance.createStyleSheet('foo', 'default');

      expect(instance.transformStyles(styleSheet.button)).toMatchSnapshot();
    });

    it('supports CSS modules', () => {
      instance = new CSSModulesAesthetic();
      instance.registerTheme('default', {});
      instance.setStyles('foo', () => ({ button: 'button' }));

      const styleSheet = instance.createStyleSheet('foo', 'default');

      expect(instance.transformStyles(styleSheet.button)).toMatchSnapshot();
    });

    it('supports Fela', () => {
      instance = new FelaAesthetic(
        createRenderer({
          plugins: [...webPreset],
        }),
      );
      instance.registerTheme('default', {});
      instance.setStyles('foo', () => SYNTAX_UNIFIED_LOCAL_FULL);

      const styleSheet = instance.createStyleSheet('foo', 'default');

      expect(instance.transformStyles(styleSheet.button)).toMatchSnapshot();
    });

    it('supports JSS', () => {
      const jss = create();
      jss.setup(preset());

      instance = new JSSAesthetic(jss);
      instance.registerTheme('default', {});
      instance.setStyles('foo', () => SYNTAX_UNIFIED_LOCAL_FULL);

      const styleSheet = instance.createStyleSheet('foo', 'default');

      expect(instance.transformStyles(styleSheet.button)).toMatchSnapshot();
    });

    it('supports TypeStyle', () => {
      instance = new TypeStyleAesthetic();
      instance.registerTheme('default', {});
      instance.setStyles('foo', () => SYNTAX_UNIFIED_LOCAL_FULL);

      const styleSheet = instance.createStyleSheet('foo', 'default');

      expect(instance.transformStyles(styleSheet.button)).toMatchSnapshot();
    });
  });
});
