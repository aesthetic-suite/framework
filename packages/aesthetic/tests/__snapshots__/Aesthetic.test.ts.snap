// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aesthetic adapters supports Aphrodite 1`] = `"button_17ap0k4"`;

exports[`Aesthetic adapters supports CSS modules 1`] = `"button"`;

exports[`Aesthetic adapters supports Fela 1`] = `"a b c d e f g h i j k l m n o p q r s t u v w x"`;

exports[`Aesthetic adapters supports JSS 1`] = `"foo-button-0-1-1"`;

exports[`Aesthetic adapters supports TypeStyle 1`] = `"f1lp6j3"`;

exports[`Aesthetic adapters supports class names 1`] = `"button"`;

exports[`Aesthetic extendTheme() errors if the parent theme doesnt exist 1`] = `"Theme \\"foo\\" does not exist."`;

exports[`Aesthetic getStyles() errors if no theme 1`] = `"Theme \\"unknown\\" does not exist."`;

exports[`Aesthetic getTheme() errors if the theme doesnt exist 1`] = `"Theme \\"foo\\" does not exist."`;

exports[`Aesthetic registerTheme() errors if a theme name has been used 1`] = `"Theme \\"foo\\" already exists."`;

exports[`Aesthetic registerTheme() errors if a theme style is not an object 1`] = `"Theme \\"foo\\" must be a style object."`;

exports[`Aesthetic registerTheme() errors if global styles is not an object 1`] = `"Global styles for \\"foo\\" must be a function."`;

exports[`Aesthetic setStyles() errors if extended and style names match 1`] = `"Cannot extend from \\"foo\\" as those styles do not exist."`;

exports[`Aesthetic setStyles() errors if extended styles do not exist 1`] = `"Cannot extend from \\"parent\\" as those styles do not exist."`;

exports[`Aesthetic setStyles() errors if styles are not a function 1`] = `"Styles defined for \\"foo\\" must be a function."`;

exports[`Aesthetic setStyles() errors if styles are not a function 2`] = `"Styles defined for \\"foo\\" must be a function."`;

exports[`Aesthetic setStyles() errors if styles are not a function 3`] = `"Styles defined for \\"foo\\" must be a function."`;

exports[`Aesthetic setStyles() errors if styles are not a function 4`] = `"Styles defined for \\"foo\\" must be a function."`;

exports[`Aesthetic setStyles() errors if styles have been set 1`] = `"Styles have already been set for \\"foo\\"."`;

exports[`Aesthetic setStyles() errors if styles try to extend from itself 1`] = `"Styles have already been set for \\"foo\\"."`;

exports[`Aesthetic transformStyles() errors for invalid value 1`] = `"Unsupported style type to transform."`;
